cmd/dockerd/daemon.go

```diff
@@ -291,34 +291,39 @@ func newRouterOptions(config *config.Config, d *daemon.Daemon) (routerOptions, e
 		return opts, err
 	}
 	cgroupParent := newCgroupParent(config)
-	bk, err := buildkit.New(buildkit.Opt{
-		SessionManager:      sm,
-		Root:                filepath.Join(config.Root, "buildkit"),
-		Dist:                d.DistributionServices(),
-		NetworkController:   d.NetworkController(),
-		DefaultCgroupParent: cgroupParent,
-		RegistryHosts:       d.RegistryHosts(),
-		BuilderConfig:       config.Builder,
-		Rootless:            d.Rootless(),
-		IdentityMapping:     d.IdentityMapping(),
-		DNSConfig:           config.DNSConfig,
-		ApparmorProfile:     daemon.DefaultApparmorProfile(),
-	})
-	if err != nil {
-		return opts, err
-	}
-
-	bb, err := buildbackend.NewBackend(d.ImageService(), manager, bk, d.EventsService)
-	if err != nil {
-		return opts, errors.Wrap(err, "failed to create buildmanager")
-	}
-	return routerOptions{
+	ro := routerOptions{
 		sessionManager: sm,
-		buildBackend:   bb,
-		buildkit:       bk,
 		features:       d.Features(),
 		daemon:         d,
-	}, nil
+	}
+	if !d.UsesSnapshotter() {
+		bk, err := buildkit.New(buildkit.Opt{
+			SessionManager:      sm,
+			Root:                filepath.Join(config.Root, "buildkit"),
+			Dist:                d.DistributionServices(),
```

Comment:
From what I can tell, builder-next needs to be skipped when the containerd snapshotter is used because `buildkit.New` would otherwise return an error due to `d.DistributionServices()` returning a zero-valued struct? Have you considered testing for that directly so `d.UsesSnapshotter()` would not have to be exported and buildkit would be reenabled automatically once `func (cs *ImageService) DistributionServices()` is implemented?
```suggestion
	if dist := d.DistributionServices(); dist != images.DistributionServices{} {
		bk, err := buildkit.New(buildkit.Opt{
			SessionManager:      sm,
			Root:                filepath.Join(config.Root, "buildkit"),
			Dist:                dist,
```

Reply:
I kept this one for now, as it felt slightly more explicit (but I agree it'd be nice to un-export the function as soon as we can) but let me know if you feel strongly about this.

Reply:
Nope, I don't feel strongly about it